pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,506 @@
+				Error::<T>::RewardIdNotRegister
+			);
+
+			// check if rewards can be unlocked at current block
+			if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+				ensure!(
+					reward_info.start_block.saturated_into::<u128>() <=
+						<frame_system::Pallet<T>>::block_number().saturated_into::<u128>(),
+					Error::<T>::RewardsCannotBeUnlockYet
+				);
+			} else {
+				//sanity check
+				return Err(Error::<T>::RewardIdNotRegister.into())
+			}
+
+			//check if user has already initialize the reward
+			ensure!(
+				!<Distributor<T>>::contains_key(reward_id, &user),
+				Error::<T>::RewardsAlreadyInitialized
+			);
+
+			let account_in_vec: Vec<u8> = T::AccountId::encode(&user);
+			#[allow(clippy::borrow_interior_mutable_const)]
+			#[allow(clippy::declare_interior_mutable_const)]
+			//get info of user from pre defined hash map
+			if let Some((total_rewards_in_pdex, initial_rewards_claimable, factor)) =
+				crowdloan_rewardees::HASHMAP.get(&account_in_vec)
+			{
+				//get reward info
+				if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+					if *total_rewards_in_pdex > MIN_REWARDS_CLAIMABLE_AMOUNT {
+						//initialize reward info struct
+						let reward_info = RewardInfoForAccount {
+							total_reward_amount: (*total_rewards_in_pdex).saturated_into(),
+							claim_amount: 0_u128.saturated_into(),
+							is_initial_rewards_claimed: false,
+							is_initialized: false,
+							lock_id: REWARDS_LOCK_ID,
+							last_block_rewards_claim: reward_info.start_block,
+							initial_rewards_claimable: (*initial_rewards_claimable)
+								.saturated_into(),
+							factor: (*factor).saturated_into(),
+						};
+						//insert reward info into storage
+						<Distributor<T>>::insert(reward_id, user.clone(), reward_info);
+					} else {
+						return Err(Error::<T>::AmountToLowtoInitializeRewards.into())
+					}
+				} else {
+					//sanity check
+					return Err(Error::<T>::RewardIdNotRegister.into())
+				}
+			} else {
+				return Err(Error::<T>::UserNotEligible.into())
+			}
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
+			<Distributor<T>>::mutate(reward_id, user.clone(), |user_reward_info| {
+				if let Some(user_reward_info) = user_reward_info {
+					//check if user already unlocked the rewards
+					//sanity check: if user is calling initialize_for the second time it will fail
+					// in the above ensure, if user is calling initialize for the first time the
+					// flag is set to false by default
+					ensure!(
+						!user_reward_info.is_initialized,
+						Error::<T>::RewardsAlreadyInitialized
+					);
+					//transfer funds from pallet account to users account
+					ensure!(
+						Self::transfer_pdex_rewards(
+							&Self::get_pallet_account(),
+							&user,
+							user_reward_info.total_reward_amount
+						)
+						.is_ok(),
+						Error::<T>::TransferFailed
+					);
```

Comment:
this logic is shady at best. you did check on line 196, then inserted, then check again and then mutate what you JUST inserted. 2 reads and 2 writes for no reason at all. rework into proper logic